src/puppetlabs/dujour/version_check.clj

```diff
@@ -0,0 +1,116 @@
+(ns puppetlabs.dujour.version-check
+  (:require [clojure.tools.logging :as log]
+            [schema.core :as schema]
+            [ring.util.codec :as ring-codec]
+            [puppetlabs.http.client.sync :as client]
+            [cheshire.core :as json]
+            [trptcolin.versioneer.core :as version]
+            [slingshot.slingshot :as sling]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Constants
+
+(def default-group-id "puppetlabs.packages")
+(def default-update-server-url "http://updates.puppetlabs.com")
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Schemas
+
+(def ProductCoords
+  {:group-id schema/Str
+   :artifact-id schema/Str})
+
+(def ProductName
+  (schema/conditional
+    #(string? %) schema/Str
+    #(map? %) ProductCoords))
+
+(def UpdateInfo
+  {:version schema/Str
+   :newer   schema/Bool
+   :link    schema/Str
+   :product schema/Str
+   (schema/optional-key :message) schema/Str})
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Private
+
+(defn schema-match?
+  [schema obj]
+  (nil? (schema/check schema obj)))
+
+(schema/defn get-coords :- ProductCoords
+  [product-name :- ProductName]
+  (condp schema-match? product-name
+    schema/Str {:group-id default-group-id
+                :artifact-id product-name}
+    ProductCoords product-name))
+
+(schema/defn version*
+  "Get the version number of this installation."
+  [group-id artifact-id]
+  {:post [(string? %)]}
+  (version/get-version group-id artifact-id))
+
+(def version
+  "Get the version number of this installation."
+  (memoize version*))
+
+(schema/defn ^:always-validate update-info :- (schema/maybe UpdateInfo)
+  "Make a request to the puppetlabs server to determine the latest available
+  version. Returns the JSON object received from the server, which
+  is expected to be a map containing keys `:version`, `:newer`, and `:link`.
+  Returns `nil` if the request does not succeed for some reason."
+  [product-name :- ProductName
+   update-server-url :- schema/Str]
+  (let [{:keys [group-id artifact-id]} (get-coords product-name)
+        current-version (version group-id artifact-id)
+        version-data {:version current-version}
+        query-string (ring-codec/form-encode version-data)
+        url (format "%s?product=%s&group=%s&%s" update-server-url artifact-id group-id query-string)
+        {:keys [status body] :as resp} (client/get url
```

Comment:
These are questions I've raised in the past, and the impression that I've gotten is that we need to put some more thought into what sorts of things we'd feel OK about doing w/o having an opt-in mechanism in place first... so, yeah, I think we need to punt that for now.

I do think it might be worth going ahead and making this function accept a map as the first argument, with product name being a key in the map.  Or adding another argument that was a map of additional values.  The service could be written to go ahead and serialize the key/value pairs from the map into the request, so that consumers of this library can send up additional data as they see fit... but I don't think we should make any changes here or now that actually involve sending any additional data.